lib/engine/game/g_system18/map_gotland_customization.rb

```diff
@@ -0,0 +1,535 @@
+              ability.float_cost = new_cost
+              ability.description = "Float cost: #{new_cost}"
+              ability.desc_detail = "This corporation requires #{new_cost} to float"
+            end
+          end
+        end
+
+        def can_par?(corporation, player)
+          player_president_shares = corporations.count { |corp| corp.president?(player) }
+          return false if player_president_shares >= 4
+
+          super
+        end
+
+        def float_corporation(corporation)
+          player = corporation.owner
+
+          ability = corporation.all_abilities.find do |a|
+            a.is_a?(Engine::Ability::FloatCost) && a.float_cost && a.float_cost.positive?
+          end
+          if ability
+            float_cost = ability.float_cost
+            if player.cash < float_cost
+              raise GameError,
+                    "#{player.name} does not have enough cash to float #{corporation.name}. "
+            end
+            player.spend(float_cost, @bank)
+            @log << "#{player.name} pays #{format_currency(float_cost)} to float #{corporation.name}"
+            corporation.remove_ability(ability)
+          end
+          floated_index = @corporation_float_order.index { |c| c[:sym] == corporation.id }
+          reduce_float_costs(floated_index)
+          super
+          # Track newly floated corporations for the current stock round
+          @round.track_newly_floated(corporation) if @round.is_a?(GSystem18::Round::GotlandStock)
+        end
+
+        # ----- Rival share randomizer -----
+        def rival_share_randomizer
+          @log << '-- Rival share randomizer --'
+          dice_num = (rand.to_i % 6) + 1
+          @log << "Rivals share random action: #{dice_num} - #{CHOICES[dice_num]}"
+          case dice_num
+          when 1
+            handle_sell(rival, :all)
+          when 2
+            handle_sell(rival, 1)
+          when 3
+            @log << 'Rival passes'
+          when 4, 5, 6
+            handle_buy(rival, dice_num - 3)
+          end
+        end
+
+        def handle_buy(rival, count)
+          bundles = find_buyable_bundles(rival, count)
+          return @log << 'No available shares to buy' if bundles.empty?
+
+          bundles.each { |bundle| process_buy_shares(bundle) }
+        end
+
+        def find_buyable_bundles(_entity, count)
+          bundles = []
+          corporations.select(&:floated?).sort_by { |corp| -(corp.share_price&.price || 0) }.each do |corp|
+            break if bundles.size >= count
+
+            share = corp.treasury_shares&.first || share_pool.shares_by_corporation[corp]&.first
+            bundles << share.to_bundle if share
+          end
+          bundles
+        end
+
+        def process_buy_shares(bundle)
+          @log << "Rival buys one share from #{bundle.corporation.name}"
+          share_pool.buy_shares(rival, bundle, exchange: true)
+        end
```

Comment:
process_* methods need to be in step/ files, not at this level. You need to create your own step/buy_sell_shares.rb that is gotland-specific.